hail/python/hailtop/aiotools/fs.py

```diff
@@ -665,9 +672,12 @@ async def copy_source(srcentry):
 
             await self._copy_file_multi_part(sema, source_report, srcfile, await srcentry.status(), url_join(full_dest, relsrcfile), return_exceptions)
 
-        await bounded_gather2(sema, *[
-            copy_source(srcentry)
-            async for srcentry in srcentries], cancel_on_error=True)
+        if srcentries is None:
```

Comment:
I'm uneasy with the empty directory creation. Even if we handle this use case where a top level directory is an empty file/directory, there's still downstream empty file/directory being missed completely and we won't make those empty directories on the local file system where mkdir is not a noop.

For example,

path/foo/ is the only match

versus.

path/bar/foo/ and the query is path/bar/.